light-client/src/consensus/consensus_client.rs

```diff
@@ -588,6 +475,124 @@ impl<R: ConsensusRpc> ConsensusLightClient<R> {
     }
 }
 
+// implements checks from validate_light_client_update and process_light_client_update in the
+// specification
+pub fn verify_generic_update(
+    store: &LightClientStore,
+    update: &GenericUpdate,
+    expected_slot: u64,
+    genesis_root: &[u8],
+    fork_version: &[u8],
+) -> Result<()> {
+    let bits = get_bits(&update.sync_aggregate.sync_committee_bits);
+    if bits == 0 {
+        return Err(ConsensusError::InsufficientParticipation.into());
+    }
+
+    let update_finalized_slot = update.finalized_header.clone().unwrap_or_default().slot;
+    let valid_time = expected_slot >= update.signature_slot
+        && update.signature_slot > update.attested_header.slot
+        && update.attested_header.slot >= update_finalized_slot;
+
+    if !valid_time {
+        return Err(ConsensusError::InvalidTimestamp.into());
+    }
+
+    let store_period = calc_sync_period(store.finalized_header.slot);
+    let update_sig_period = calc_sync_period(update.signature_slot);
+    let valid_period = if store.next_sync_committee.is_some() {
+        update_sig_period == store_period || update_sig_period == store_period + 1
+    } else {
+        update_sig_period == store_period
+    };
+
+    if !valid_period {
+        return Err(ConsensusError::InvalidPeriod.into());
+    }
+
+    let update_attested_period = calc_sync_period(update.attested_header.slot);
+    let update_has_next_committee = store.next_sync_committee.is_none()
+        && update.next_sync_committee.is_some()
+        && update_attested_period == store_period;
+
+    if update.attested_header.slot <= store.finalized_header.slot && !update_has_next_committee {
+        return Err(ConsensusError::NotRelevant.into());
+    }
+
+    if update.finalized_header.is_some() && update.finality_branch.is_some() {
+        let is_valid = is_finality_proof_valid(
+            &update.attested_header,
+            &mut update
+                .finalized_header
+                .clone()
+                .expect("we know that this is `Some`"),
+            &update
+                .finality_branch
+                .clone()
+                .expect("we know that this is `Some`"),
+        );
+
+        if !is_valid {
+            return Err(ConsensusError::InvalidFinalityProof.into());
+        }
+    }
+
+    if update.next_sync_committee.is_some() && update.next_sync_committee_branch.is_some() {
+        let is_valid = is_next_committee_proof_valid(
+            &update.attested_header,
+            &mut update
+                .next_sync_committee
+                .clone()
+                .expect("we know that this is `Some`"),
+            &update
+                .next_sync_committee_branch
+                .clone()
+                .expect("we know that this is `Some`"),
+        );
+
+        if !is_valid {
+            return Err(ConsensusError::InvalidNextSyncCommitteeProof.into());
+        }
```

Comment:
```suggestion
        ensure!(is_valid, ConsensusError::InvalidNextSyncCommitteeProof.into());
```
nit: I think this is cleaner